// @lalamove/lalamove-js/dist/payload/orderPayloadBuilder.js
"use strict";
var __importDefault = (this && this.__importDefault) || function (mod) {
    return (mod && mod.__esModule) ? mod : { "default": mod };
};
Object.defineProperty(exports, "__esModule", { value: true });
const orderPayload_1 = __importDefault(require("./orderPayload"));
class OrderPayloadBuilder {
    static orderPayload() {
        return new OrderPayloadBuilder();
    }
    withQuotationID(quotationId) {
        this.quotationId = quotationId;
        return this;
    }
    withSender(sender) {
        this.sender = sender;
        return this;
    }
    withRecipients(recipients) {
        this.recipients = recipients;
        return this;
    }
    withIsPODEnabled(isPodEnabled) {
        this.isPODEnabled = isPodEnabled;
        return this;
    }
    withIsRecipientSmsEnabled(isRecipientSMSEnabled) {
        this.isRecipientSMSEnabled = isRecipientSMSEnabled;
        return this;
    }
    withPartner(partner) {
        this.partner = partner;
        return this;
    }
    withMetadata(metadata) {
        this.metadata = metadata;
        return this;
    }
    build() {
        return new orderPayload_1.default(this);
    }
}
exports.default = OrderPayloadBuilder;
